design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,291 @@
+
+### Varargs
+
+Varargs boils down to supporting native Java arrays as parameters and ensuring that component types 
+are resolved properly. Anytime a method is registered with variable args (e.g. `String... args`) we 
+will register a corresponding function with the wrapping array type (e.g. `String[]`). At runtime, 
+if no function matches the explicit parameter types, we will fallback to search for any array-type 
+parameter method declaration of the same type iff all parameters are of a single type. If any
+parameter is null, it will be considered valid for any vararg declaration.
+
+This proposal does not cover supporting `Object` as a parameter to a method in order to allow for 
+"generic variable argument" UDFs, but supporting it can be an extension.
+
+```java
+@Udf("returns a sublist of 'list' t")
+public int sum(@UdfParameter final int... args) {
+  return Arrays.stream(args).sum();
+}
+```
+
+### Complex Aggregation
+
+We will allow users to supply an additional method `VR export(A agg)`. This method will be taken
+as the behavioral parameter to a `mapValues` task that will be applied after the `aggregate` task 
+in the generated KStreams topology. To support this in a backwards compatible fashion, we will 
+introduce a new interface `Exportable`, that UDFs may implement. Only UDAFs that implement this 
+interface will have the `mapValues` stage applied to them.
+
+```java
+class AverageUdaf implements Udaf<Long, Struct>, Exportable<Struct, Long> {
+  
+  private static final Schema SUM_SCHEMA = 
+      SchemaBuilder.struct()
+                  .field("sum", Schema.INT64_SCHEMA)
+                  .field("count", Schema.INT64_SCHEMA).build();
+  
+  @Override
+  public Struct initialize() {
+    return new Struct(SCHEMA);
+  }
+  
+  @Override
+  public Struct aggregate(final Long val, final Struct agg) {
+    agg.put("sum", agg.get("sum") + val);
+    agg.put("count", agg.get("count") + 1);
+    return agg;
+  }
+  
+  @Override
+  public Struct merge(final Struct agg1, final Struct agg2) {
+    agg1.put("sum", agg1.get("sum") + agg2.get("sum"));
+    agg1.put("count", agg1.get("count") + agg2.get("count"));
+    return agg1;
+  }
+ 
+  /**
+  * This is the new method that overrides {@code Exportable#export(A agg)} (note that
+  * {@code Exportable<A,VR>} is defined with type parameters for the input and output
+  * of this method.
+  */
+  @Override
+  public Integer export(final Struct agg) {
+    return agg.getInt64("sum") / agg.getInt64("count");
+  }
+  
+}
+```
+
+**NOTE:** It should be possible to support arbitrary Java objects for aggregation without much
+extra work. For the scope of this KLIP, however, supporting just KSQL types (including `Struct`) 
+should suffice.
+
+#### Alternative
+
+If we expect UDAFs to commonly require exportable functionality, then we can make this change in a
+backwards incompatible change and introduce the `export` (or `terminate`) method into the UDAF
+interface directly.
```

Comment:
i don't think this would be the end of the world from a UX perspective, just needs documenting properly in the release notes. We could obviously convert over the in-built UDAFs as we implement this. I do believe the majority of non-trivial UDAFs will need to avail themselves of this new capability.

Reply:
@MichaelDrogalis thoughts?